net_util: Simplify TX/RX queue handling

sboeuf opened the pull request:
The main idea behind this commit is to remove all the complexity
associated with TX/RX handling for virtio-net. By using writev() and
readv() syscalls, we could get rid of intermediate buffers for both
queues.

The complexity regarding the TAP registration has been simplified as
well. The RX queue is only processed when some data are ready to be
read from TAP. The event related to the RX queue getting more
descriptors only serves the purpose to register the TAP file if it's not
already.

With all these simplifications, the code is more readable but more
performant as well. We can see an improvement of 3Gbits/s for a single
queue device.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

rbradford commented on net_util/src/queue_pair.rs:
```diff
@@ -128,63 +107,64 @@ impl RxVirtio {
+                        tap.as_raw_fd() as libc::c_int,
+                        iovecs.as_ptr() as *const libc::iovec,
+                        iovecs.len() as libc::c_int,
+                    )
```
One of the reasons why there is an intermediate buffer is because when you ask to read from a TAP device you must read the whole frame in one go; if you read only a fraction then the remainder is "lost". I don't see how you handle that situation.

rbradford replied:
https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1265#issuecomment-638781834

rbradford replied:
Or is the kernel providing a descriptor chain of <header size> <MTU size> <header size> <MTU size> ... ?

sboeuf replied:
The TAP always provide whole frames, the problem was coming from the way we were reading/writing each time into a descriptor, which was not matching the size of the frame.
By using vectorized read/write, we solve the problem as the whole frame can be read/written in one go.

rbradford replied:
Okay, cool. What did you have to fix for the CI to pass?

sboeuf replied:
I was missing the seccomp filters.

rbradford commented:
In your commit message you should refer to the improvement as a percentage; 3GiB/s doesn't tell us much.

rbradford commented:
@sboeuf I tried your branch I see this for every ping packet i send from host to guest

```cloud-hypervisor: 144.209600915s: <_net0_qp0> INFO:net_util/src/queue_pair.rs:146 -- net: rx: failed reading from tap: Resource temporarily unavailable (os error 11)```

sboeuf commented:
@rbradford I've been reading the virtio spec, we never did a single thing correctly, and this PR is not doing things better. I'm gonna have to re-write `virtio-net`.

rbradford commented on net_util/src/queue_pair.rs:
```diff
@@ -128,63 +107,64 @@ impl RxVirtio {
+                if result < 0 {
+                    // This is not considered an error as we might try to read
+                    // too much from the TAP. If that's the case, we simply
+                    // want to stop processing the descriptors.
```
You should only ignore the case where the error is EAGAIN. All others should be reported and handled upwards.

rbradford commented:
master:

```
--- 192.168.249.2 ping statistics ---
100 packets transmitted, 100 received, 0% packet loss, time 101269ms
rtt min/avg/max/mdev = 0.485/0.653/1.424/0.110 ms
```
This PR:

```
--- 192.168.249.2 ping statistics ---
100 packets transmitted, 100 received, 0% packet loss, time 99782ms
rtt min/avg/max/mdev = 0.473/0.922/1.293/0.132 ms
```

Do you have any thoughts about the increased latency, ~40% ?

rbradford commented:
> master:
> 
> ```
> --- 192.168.249.2 ping statistics ---
> 100 packets transmitted, 100 received, 0% packet loss, time 101269ms
> rtt min/avg/max/mdev = 0.485/0.653/1.424/0.110 ms
> ```
> This PR:
> 
> ```
> --- 192.168.249.2 ping statistics ---
> 100 packets transmitted, 100 received, 0% packet loss, time 99782ms
> rtt min/avg/max/mdev = 0.473/0.922/1.293/0.132 ms
> ```
> 
> Do you have any thoughts about the increased latency, ~40% ?

Oops, I realised I did this with a debug build so the data isn't so useful!

sboeuf commented:
@rbradford I've addressed all comments.
About latency, I measured based on a release build (using `ethr` tool). It's around 50us for both master and this PR, so no differences. QEMU is more around 100us, but it has better throughput.
